### PR TITLE
Add support for PWA

### PR DESCRIPTION
**Pull request by SleeplessOne1917**
This was originally meant to address #996, however I decided to set up the service worker so that it would work as a PWA properly on multiple platforms.

### CODE REVIEW

File: dev.dockerfile [hunk 1]
```diff
@@ -3,9 +3,13 @@ RUN apk update && apk add curl yarn python3 build-base gcc wget git --no-cache
 
 WORKDIR /usr/src/app
 
+ENV npm_config_target_arch=x64
+ENV npm_config_target_platform=linux
+ENV npm_config_target_libc=musl
+
 # Cache deps
 COPY package.json yarn.lock ./
-RUN yarn install --ignore-scripts --prefer-offline --pure-lockfile
```
Comment (SleeplessOne1917) [inline dev.dockerfile:8]: This and the environment variables are required to make the image processing library work. See https://github.com/lovell/sharp/issues/3663.

File: src/shared/env.ts
Comment (SleeplessOne1917) [inline src/shared/env.ts:1]: I was getting constants meant for the client on the server. Changing them to functions fixed it.

File: src/server/index.tsx [hunk 2]
```diff
@@ -267,3 +301,63 @@ function removeParam(url: string, parameter: string): string {
...
+
+const iconSizes = [72, 96, 128, 144, 152, 192, 384, 512];
+const defaultLogoPathDirectory = path.join(
+  process.cwd(),
+  "dist",
+  "assets",
+  "icons"
+);
+
+export async function generateManifestBase64(site: Site) {
```
Comment (dessalines) [inline src/server/index.tsx:313]: We should probably re-organize some of this into separate files... not necessary for this PR tho.

File: src/shared/env.ts [hunk 3]
```diff
@@ -2,49 +2,53 @@ import { isBrowser } from "./utils";
...
-export const httpBaseInternal = `http://${host}`; // Don't use secure here
-export const httpBase = `http${secure}://${host}`;
-export const wsUriBase = `ws${secure}://${wsHost}`;
-export const wsUri = `${wsUriBase}/api/v3/ws`;
-export const isHttps = secure.endsWith("s");
-
-console.log(`httpbase: ${httpBase}`);
-console.log(`wsUri: ${wsUri}`);
-console.log(`isHttps: ${isHttps}`);
+const getInternalHost = () =>
```
Comment (dessalines) [inline src/shared/env.ts:5]: Functions, not closures. Change all of these awful javascript fake functions into real ones.

File: src/server/index.tsx [hunk 4]
```diff
@@ -267,3 +301,63 @@ function removeParam(url: string, parameter: string): string {
...
+
+  const manifest = {
+    name: site.name,
+    description: site.description ?? "A link aggregator for the fediverse",
+    start_url: url,
+    scope: url,
+    display: "standalone",
+    id: "/",
+    background_color: "#222222",
+    theme_color: "#222222",
```
Comment (SleeplessOne1917) [inline src/server/index.tsx:329]: I wonder if theme_color and background_color should be configurable by instance admins?

**Review by dessalines — APPROVED** [review]

File: src/server/index.tsx (hunk 4 above)
Comment (dessalines) [reply src/server/index.tsx:329]: mmm.... lets not do it until someone asks for it lol. Plus I think that would require a local site setting in the database, or some image processing... pry overkill.

**Pull request merged**